src/avram/associations/has_many.cr

```diff
@@ -69,6 +61,31 @@ module Avram::Associations::HasMany
     end
   end
 
+  private macro define_has_many_through_base_query(assoc_name, model, through)
+    class BaseQuery
+      def preload_{{ assoc_name }}
+        preload_{{ assoc_name }}({{ model }}::BaseQuery.new)
+      end
+
+      def preload_{{ assoc_name }}
+        modified_query = yield {{ model }}::BaseQuery.new
+        preload_{{ assoc_name }}(modified_query)
+      end
+
+      def preload_{{ assoc_name }}(preload_query : {{ model }}::BaseQuery)
+        preload_{{ through.first.id }} do |through_query|
+          through_query.preload_{{ through[1].id }}(preload_query)
+        end
```

Comment:
We force the preload of the through association and in the query we tell that association to preload the next one.

This is reversing the way we were doing it before. This is required because we don't know the method names in the old direction, we were making a guess using the foreign key which doesn't handle both the has many through a belongs to and a has many through a has many scenarios.

I _think_ this ends up making more sql calls and could potentially be N+1 but I'm not sure.